### PR TITLE
SNOW-1331032 add retrieve func defaults from source

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1331032

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Splits original PR #1509.

This PR adds a new function `retrieve_func_defaults_from_source` which extracts default values from function registration when registrining sproc, udf, udtf from a file. The function is similar to `retrieve_func_defaults_from_source` and parses the AST of the given file to extract the default values.

### CODE REVIEW

**Review by sfc-gh-yuwang — CHANGES_REQUESTED** [review]
Seems like a few lines are not covered by the test, can you add them?

**Comment by sfc-gh-aalam** [comment]
> Seems like a few lines are not covered by the test, can you add them?

I added more tests. Just a note that the full PR has coverage satisfied: https://github.com/snowflakedb/snowpark-python/pull/1509

**Review by sfc-gh-yuwang — APPROVED** [review]

File: tests/unit/test_types.py [hunk 1]
```diff
@@ -586,6 +587,99 @@ def test_decimal_regular_expression(decimal_word):
...
+    if test_from_class:
+        source = f"""
+class {class_name}:
+    def {func_name}(self, x, y {datatype_str} = {annotated_value}) -> None:
+        return None
+"""
+    else:
+        source = f"""
+def {func_name}(x, y {datatype_str} = {annotated_value}) -> None:
+    return None
```
Comment (sfc-gh-aling) [inline tests/unit/test_types.py:671]: nit: have you tested multiple default values? `def func(a, b=xx, c=xxx, g=xx)`
it doesn't need to be complicated, but just testing multiple kwargs also works.

**Review by sfc-gh-aling — APPROVED** [review]

File: tests/unit/test_types.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply tests/unit/test_types.py:671]: I have it as an integration test over here: https://github.com/snowflakedb/snowpark-python/pull/1509/files#r1696077946

**Comment by sfc-gh-aalam** [comment]
https://ci-dev-142.int.snowflakecomputing.com/job/PythonStoredProcBuildPrecommitTest/1233/

**Pull request merged**